Radix sort

rchillyard opened the pull request:
This PR is to enable radix sort in Husky sort.

rchillyard committed:
Testing against (object) DualPivotQuicksort which is implemented in PureDualPivotQuicksort

rchillyard committed:
Added minimum repetitions for benchmarks

rchillyard committed:
Purely cosmetic changes.

rchillyard committed:
Implemented RadixSort for Integers

rchillyard committed:
Added code for Radix and StringSorts

MSDStringSort: Implemented sort but vulnerable to stray characters out of 0-255 range.
HuskySortBenchmark: added benchmark for MSDStringSort
RadixSortTest: tests ignored for now.

rchillyard committed:
Created new package radix

Moved classes to radix package.
Added inchoate Alphabet class.

rchillyard committed:
Introduced Alphabet

MSDStringSort now reserves space for characters out of the ASCII range.
There is no provision for growing that space and for now, we don't clear the space before a second run (but I don't think that ever happens in HuskySortBenchmark).

rchillyard committed:
A couple of improvements to MSDStringSort

Alphabet: added counts() method;
Alphabet and MSDStringSort: added reset() method;
HuskySortBenchmark: call reset();

rchillyard committed:
There are still some problems with very large arrays using unicode strings, but this is adequate for benchmarking relatively small array (up to 1,000,000).

rchillyard committed:
MSDStringSort continuing

Unit tests for same are ignored.
Still much work to be done.

rchillyard committed:
Interim commit for collator-based encoders

Added encoding for byte arrays and use the CollatorKey to get byte array for a given string.

NiftyMule commented:
Hi professor, found a small bug for huskySort sorting Chinese in pinyin order:

[PureHuskySort.java#L52](https://github.com/rchillyard/The-repository-formerly-known-as/blob/a0c878038f99e199ea4226c45f68a69b0837b6c6/src/main/java/edu/neu/coe/huskySort/sort/huskySort/PureHuskySort.java#L52), here the second pass to fix remaining inversions will need to use collactor.compare() if not sorting in unicode default order :octocat:

rchillyard committed:
Fixed the PureHuskySort code so that, if the encoder uses a Collator, then the final Arrays.sort will also use the collator.

rchillyard committed:
Fixed problem in SequenceEncoder_Collator whereby huskyEncode was not implemented;
Fixed problem in HuskyCoder (for byte array) where husky code was negative when it should have been positive.

rchillyard commented:
> Hi professor, found a small bug for huskySort sorting Chinese in pinyin order:
> 
> [PureHuskySort.java#L52](https://github.com/rchillyard/The-repository-formerly-known-as/blob/a0c878038f99e199ea4226c45f68a69b0837b6c6/src/main/java/edu/neu/coe/huskySort/sort/huskySort/PureHuskySort.java#L52), here the second pass to fix remaining inversions will need to use collactor.compare() if not sorting in unicode default order :octocat:

This has now been fixed. Maybe you'd like to create an issue(s) which covers this aspect and also the problem of the negative long numbers. I can then show this PR as fixing them.

NiftyMule commented:
Issues created! [#53] and [#54]

rchillyard committed:
Some changes in HuskySortBenchmark (and config)

Warning: one unit test failure: HuskySortBenchmark null pointer exception
Use collator for huskyCoder.
I don't know if this is a good idea but I need to get it pushed to gitHub.

rchillyard committed:
Fixed problem with NullPointerException for ordinary Strings.

lijiaqi04 committed:
chinese sort

lijiaqi04 committed:
stuck with mergeHuskySortTest

lijiaqi04 committed:
modify the test set

rchillyard committed:
added pinyin4j to pom.xml
